### ISSUE TITLE
Service resynced every ~2mins, causes consul index to grow

### ISSUE THREAD

**Issue by duanxuelin** [hashicorp/consul#4960]
We are running agent in a k8s pod. After register a service, the agent syned service continuously。
```
    2018/11/15 02:04:41 [INFO] agent: Synced node info
    2018/11/15 02:04:46 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:04:46 [INFO] agent: Synced check "service:54ebba0c-841d-4d6a-8f09-734bab38d390"
==&gt; Newer Consul version available: 1.4.0 (currently running: 1.2.3)
    2018/11/15 02:06:08 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:07:46 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:09:03 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:10:34 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:11:46 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:13:34 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:14:52 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
    2018/11/15 02:16:04 [INFO] agent: Synced service "54ebba0c-841d-4d6a-8f09-734bab38d390"
```
The k8s deploymend file is like:  

```
      - image: consul:1.2.3
        imagePullPolicy: IfNotPresent
        name: consul-agent
        lifecycle:
          preStop:
            exec:
              command: ["/bin/sh", "-c", "sleep 5; consul leave -http-addr=http://$BIND_IP:8500"]
       args:
        - "agent"
        - "-bind=0.0.0.0"
        - "-retry-join=consul-server.base"
        - "-datacenter=dc"
```

**Comment by duanxuelin** [comment]
> This is by design I think.

But the growth of index makes blocking queries unusable. And this situation did not show up when I run an agent without k8s.

**Comment by banks** [comment]
@duanxuelin as Pierre said the log output is normal - that is the agent performing periodic anti-entropy sync with servers.

I assume by "But the growth of index makes blocking queries unusable." you mean that every time it syncs the service ModifiedIndex is growing which means it's constantly sending updates to blocking query watchers right?

Can you show us the check status? If checks are failing then their output gets updated on the servers every anti-entroy sync which is likely what you are seeing, but that shouldn't update the service index used to just watch the whole service unless the check is flapping between passing and critical and so the list of services available changes every time.

**Comment by duanxuelin** [comment]
@banks yes, do as you said, the change of ModifiedIndex makes my service gets update message constantly. And the check status is passing, at least the dashboard shows that. Meanwhile I do get a connection event every second with tcp check.

Here are codes with nodejs:
```
 const ret = await consul.agent.service.register({
      name: serviceName,
      id: serviceId,
      address: serviceIP,
      port: servicePort,
      check: {
        deregistercriticalserviceafter: '1m',
        tcp: `127.0.0.1:${checkPort}`,
        timeout: '3s',
        interval: '1s'
      }
    });
```


```
    const watch = consul.watch({
      method: consul.health.service,
      options: {service: serviceName, passing: true}
    });
```

**Comment by pierresouchay** [comment]
@duanxuelin

Use this script named `consul_check_services_changes.sh`:
```shell
#!/bin/bash
if test ${1:-none} = "none"
then
  echo "USAGE: $0 Consul_URL"
  echo "       Example: localhost:8500/v1/health/service/MY_SUPER_SERVICE"
  exit 1
fi
url_to_check=$1

headers=$(mktemp)
color_diff=$(which colordiff||which cdiff||echo diff)
content=$(mktemp)
index=0
while true;
do
  url="${url_to_check}?wait=10m&index=${index}&pretty=true&stale"
  curl -fs --dump-header "$headers" -o "${content}.new" "${url}" || { echo "Failed to query ${url}"; exit 1; }
  if test $index -ne 0
  then
    ${color_diff} -u "$content" "$content.new" && echo " diff: No Differences found in service"
  fi
  index=$(grep "X-Consul-Index" "$headers" | sed 's/[^0-9]*\([0-9][0-9]*\)[^0-9]*/\1/g')
  if test "${index:-not_found}" = "not_found"
  then
    # We are in a part of Consul that does not output X-Consul-Index in headers, switch to poll
    sleep 5
    index=1
  fi
  if test ${CONSUL_SLEEP_DELAY:-0} -gt 0
  then
    sleep ${CONSUL_SLEEP_DELAY}
  fi
  mv "$content.new" "$content"
  printf "X-Consul-Index: $index at $(date) \b"
done
```

And launch it against your agent this way (if your agent is remove, you can replace `localhost` by the name of any consul agent in your cluster):

`consul_check_services_changes.sh localhost:8500/v1/health/service/:service_name`

It will display the changes in real time and help you a lot (we are using it intensively)

**Comment by duanxuelin** [comment]
@banks @duanxuelin   
Now I am sure the changes of index happened when agent synced service eventhough ths check is passed. But I don't know why.
```
    2018/11/21 02:24:16 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:17 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:18 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:19 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:19 [DEBUG] agent: Skipping remote check "serfHealth" since it is managed automatically
    2018/11/21 02:24:19 [INFO] agent: Synced service "d55e9420-899f-4272-8b64-7773117babeb"
    2018/11/21 02:24:19 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" in sync
    2018/11/21 02:24:19 [DEBUG] agent: Node info in sync
    2018/11/21 02:24:19 [DEBUG] http: Request GET /v1/health/service/network-server?wait=10m&amp;index=295733&amp;pretty=true&amp;stale (1m31.939727148s) from=127.0.0.1:57946
    2018/11/21 02:24:20 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:21 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:22 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:23 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:24 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 02:24:25 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
```

**Comment by pierresouchay** [comment]
2018/11/21 02:24:19 [DEBUG] http: Request GET /v1/health/service/network-server?wait=10m&amp;index=295733&amp;pretty=true&amp;stale (1m31.939727148s) from=127.0.0.1:57946

This line shows that the index did not change for at least 1m31 and my script did block during that time

It means the sync did occur and triggered and watch

Did the script did print again a line after? 
Please display here the output of the script, but to me, it works as intended

**Comment by duanxuelin** [comment]
@pierresouchay yes,  it did print a line:  
![_f93b2af6-1977-4b63-8b7b-ea1f178e917e](https://user-images.githubusercontent.com/30339830/48836347-76a61d80-edbd-11e8-9a6c-cc4309a6640e.png)

```
    2018/11/21 10:39:28 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:39:29 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:39:30 [DEBUG] agent: Skipping remote check "serfHealth" since it is managed automatically
    2018/11/21 10:39:30 [INFO] agent: Synced service "d55e9420-899f-4272-8b64-7773117babeb"
    2018/11/21 10:39:30 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" in sync
    2018/11/21 10:39:30 [DEBUG] agent: Node info in sync
    2018/11/21 10:39:30 [DEBUG] http: Request GET /v1/health/service/network-server?wait=10m&amp;index=303063&amp;pretty=true&amp;stale (1m29.76440681s) from=127.0.0.1:40402
    2018/11/21 10:39:30 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:39:30 [DEBUG] memberlist: Initiating push/pull sync with: 10.246.3.91:8301
    2018/11/21 10:39:31 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:39:32 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
```

```
    2018/11/21 10:40:38 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:40:39 [DEBUG] agent: Skipping remote check "serfHealth" since it is managed automatically
    2018/11/21 10:40:39 [INFO] agent: Synced service "d55e9420-899f-4272-8b64-7773117babeb"
    2018/11/21 10:40:39 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" in sync
    2018/11/21 10:40:39 [DEBUG] agent: Node info in sync
    2018/11/21 10:40:39 [DEBUG] http: Request GET /v1/health/service/network-server?wait=10m&amp;index=303089&amp;pretty=true&amp;stale (1m8.386090357s) from=127.0.0.1:41212
    2018/11/21 10:40:39 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:40:40 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:40:41 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:40:42 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:40:43 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
    2018/11/21 10:40:44 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" is passing
```

The index grows every time when a sync happened. It doesn't make sense.

**Comment by banks** [comment]
@duanxuelin does your check output something different every time? If so
then every sync we will update the output of the script on the server and
that causes the update. Typically you should aim for checks to output
exactly the same thing (if the output anything) for passing state otherwise
the cluster has to do lots of extra work keeping that output in sync.

**Comment by duanxuelin** [comment]
@banks I use tcp check. The agent start a connection and I accept it, nothing more.

**Comment by banks** [comment]
Hmm so do you have a consul agent running inside your pod? If not, how does 127.0.01 in the TCP check resolve to the service in another pod? If so, I'm still confused, but it's worth noting that we don't generally recommend running in containers this way - it's preferable to deploy the agent as a daemon set so one per node and use hostIP to access it.

That said, I'm not sure why you are seeing this behaviour.

```
    2018/11/21 10:39:30 [INFO] agent: Synced service "d55e9420-899f-4272-8b64-7773117babeb"
    2018/11/21 10:39:30 [DEBUG] agent: Check "service:d55e9420-899f-4272-8b64-7773117babeb" in sync
    2018/11/21 10:39:30 [DEBUG] agent: Node info in sync
```

I should have looked close, you can see there that the check is "in sync" it's the service registration itself that is apparently not in sync.

I know you pasted code for registration already, but can you link to the JS API client you are using just in case it's doing something weird?

Also can you paste the following output (similar to the ones I requested a few messages ago) If the agent is in your pod then you'll need to run this inside the pod with kubectl exec.

`curl -s localhost:8500/v1/agent/service/d55e9420-899f-4272-8b64-7773117babeb`

(assuming your service still has that UUID as it's name/ID, if not show that for all the services actually running).

If you could paste the output of that twice - a few minutes apart (after you've seen a sync). That will hopefully help us figure out what is changing about your service (or if nothing maybe help us reproduce).

**Comment by duanxuelin** [comment]
@pierresouchay @banks 
Problem solved after I upgraded consul 1.3.0.
Thanks for help.
